proofs/proof.ml

```diff
@@ -365,34 +358,46 @@ let run_tactic env tac pr =
   let open Proofview.Notations in
   let undef sigma l = List.filter (fun g -> Evd.is_undefined sigma g) l in
   let tac =
+    Proofview.tclEVARMAP >>= fun sigma ->
+    Proofview.Unsafe.tclEVARS (Evd.push_shelf sigma) >>= fun () ->
     tac >>= fun result ->
     Proofview.tclEVARMAP >>= fun sigma ->
     (* Already solved goals are not to be counted as shelved. Nor are
       they to be marked as unresolvable. *)
     let retrieved, sigma = Evd.pop_future_goals sigma in
     let retrieved = Evd.FutureGoals.filter (Evd.is_undefined sigma) retrieved in
-    let retrieved = List.rev_append retrieved.Evd.FutureGoals.shelf (List.rev retrieved.Evd.FutureGoals.comb) in
+    let retrieved = List.rev retrieved.Evd.FutureGoals.comb in
     let sigma = Proofview.Unsafe.mark_as_goals sigma retrieved in
+    let to_shelve, sigma = Evd.pop_shelf sigma in
     Proofview.Unsafe.tclEVARS sigma >>= fun () ->
-    Proofview.tclUNIT (result,retrieved)
+    Proofview.tclUNIT (result,retrieved,to_shelve)
   in
   let { name; poly; proofview } = pr in
   let proofview = Proofview.Unsafe.push_future_goals proofview in
-  let ((result,retrieved),proofview,(status,to_shelve),info_trace) =
+  let ((result,retrieved,to_shelve),proofview,status,info_trace) =
     Proofview.apply ~name ~poly env tac proofview
   in
   let sigma = Proofview.return proofview in
   let to_shelve = undef sigma to_shelve in
-  let shelf = (undef sigma pr.shelf)@retrieved@to_shelve in
   let proofview = Proofview.Unsafe.mark_as_unresolvables proofview to_shelve in
-  { pr with proofview ; shelf },(status,info_trace),result
+  let modify = function
+    | [] -> CErrors.anomaly Pp.(str"shelf stack should not be empty")
+    | hd :: q ->
+      (hd@retrieved@to_shelve) :: q
+  in
+  let proofview = Proofview.modify_global_shelf (fun shelf -> modify (List.map (undef sigma) shelf)) proofview in
+  { pr with proofview },(status,info_trace),result
 
 (*** Commands ***)
 
 (* Remove all the goals from the shelf and adds them at the end of the
    focused goals. *)
 let unshelve p =
-  { p with proofview = Proofview.unshelve (p.shelf) (p.proofview) ; shelf = [] }
+  let sigma = Proofview.return p.proofview in
+  let shelf = Evd.shelf sigma in
+  let proofview = Proofview.unshelve shelf p.proofview in
+  let proofview = Proofview.modify_global_shelf (CList.map (fun _ -> [])) proofview in
```

Comment:
If I understand correctly, we don’t want to just set the shelf stack to [[]] here as this would maybe cause pop_shelve failures. However unshelve is a command so that should still be ok. The current implementation could allow an unshelve tactic that pops the last shelved goals and replaces it with [], and maybe even a global_unshelve tactic that clears the whole stack and turn the union of shelved goals into new goals. My question is somehow could we get an interface for well-scoped shelf manipulations at the tactic level?

Reply:
I'm working on the API just as we speak ;)

Reply:
Should be better now.